Updates to image optimization settings

SiobhyB opened the pull request:
> **Note**
> Although this PR is large, the vast majority of file changes are the result of a change to a string name, as can be viewed in https://github.com/wordpress-mobile/WordPress-Android/pull/19581/commits/452ed672ac3d147571376dc954eb907bb4f34c23

## Description

This PR changes the default setting for `Optimize Images` (found via _Me_ → _App Settings_) from off → on. We will also inform users of this new default via a popup that will display when uploading media.

The default `Maximum Image Size` has also been updated from _3000px_  to _2000px_ and the `Image Quality`'s _Medium_ value has been changed from _85%_ to _80%_. 

You can find the iOS equivalent at https://github.com/wordpress-mobile/WordPress-iOS/pull/21981.

### Included PRs

The following three PRs were merged into this one, with all of them having been separately reviewed and approved:

* https://github.com/wordpress-mobile/WordPress-Android/pull/19556
* https://github.com/wordpress-mobile/WordPress-Android/pull/19648
* https://github.com/wordpress-mobile/WordPress-Android/pull/19644

## To test

<details>
 <summary>Verify default setting displays as expected ⤵️</summary>

### Default setting should be changed if it has never been set before

* Perform a fresh installation of the app.
* Navigate to _Me → App Settings_. 
* Observe that _Optimize Images_ setting is enabled by default.
* Observe that _Maximum Image Size_ and _Image Quality_ settings have the default values as shared in the PR's description.

### Default setting should not be changed if it has been set previously

#### Set original as maximum image size
* Install an older version.
* Navigate to _Me → App Settings_. 
* Move the _Maximum Image Size_ setting to a different value than the default.
* Move the _Maximum Image Size_ setting back to the original value.
* Install the installable build from this PR.
* Navigate to _Me → App Settings_. 
* Observe that the _Optimize Images_ setting is enabled and that _Maximum Image Size_ setting remains in the previously set original value.

#### Set maximum image size
* Install an older version.
* Navigate to _Me → App Settings_. 
* Move the _Maximum Image Size_ setting to a different value than the original.
* Install the installable build from this PR.
* Navigate to _Me → App Settings_. 
* Observe that the _Optimize Images_ setting is enabled and that _Maximum Image Size_ setting remains in the previously set value.
</details>

<details>
 <summary>Resize settings should function as expected ⤵️</summary>

* Add an image to the device with dimensions above 2000px.
* Remove and install the app.
* Navigate to the Posts/Pages screen.
* Open/create a post/page.
* Add an Image block.
* Tap on the _Choose from device_ option.
* Select the image previously added with large dimensions.
* When the _Optimize Images_ pop-up is displayed, select the option to keep image optimization.
* Wait for the image upload to finish.
* Go to the browser and navigate to the Media screen of the site.
* Select the uploaded image and click on the _Edit_ button.
* Observe that dimensions displayed under the _DIMENSIONS_ section are lower than the original ones. The dimensions shouldn’t exceed the value set in _Max Image Upload Size_ setting.
* Observe that the file size under the _FILE SIZE_ section is smaller than the original file size.</details>

<details>
 <summary>Compression settings should function as expected ⤵️</summary>

- Add an image to the device with a large file size.
- Remove and install the app.
- Navigate to the Posts/Pages screen.
- Open/create a post/page.
- Add an Image block.
- Tap on the _Choose from device_ option.
- Select the image previously added with a large file size.
- When the _Optimize Images_ pop-up is displayed, select the option to keep image optimization.
- Wait for the image upload to finish.
- Go to the browser and navigate to the Media screen of the site.
- Select the uploaded image and click on the _Edit_ button.
- Observe that the file size under the _FILE SIZE_ section is smaller than the original file size.</details>

<details>
<summary>Pop-up should be displayed in specific sources ⤵️</summary>

### Media Screen
- Remove and install the app.
- Navigate to `My Site → Media`.
- Tap on ➕ button and select `Choose from My Device` option.
- Select one or multiple items.
- Tap on the `Add` button.
- Observe that the Optimize Images pop-up is displayed.

### Gutenberg Editor

**Preparation:**
- Remove and install the app.
- Navigate to `My Site → Posts` (or Pages).
- Create a post.

#### Media Blocks - From Device
- Add any of the following blocks:
  - Image block
  - Gallery block
  - Cover block
  - Media & Text block
  - Story block.
- Tap on the `Choose from device` option.
- Select one item (or multiple items when using Gallery block).
- Observe that the Optimize Images pop-up is displayed.

#### Media Blocks - Take Photo

> **Note**
> The pop up doesn't currently display when taking a photo from the Story block. This will be worked on in a future PR.

- Add any of the following blocks:
  - Image block
  - Gallery block
  - Cover block
  - Media & Text block
- Tap on the `Take a Photo` option.
- Take a photo and use it.
- Observe that the Optimize Images pop-up is displayed.

### Featured Image - From Device
- Tap on the `...` button located on the top bar.
- Tap on the `Post Settings` options.
- Navigate to the Featured Image section and tap on `Set Featured Image`.
- Tap on the `Choose from Device` option.
- Select one time.
- Observe that the Optimize Images pop-up is displayed.

### Featured Image - Take Photo
- Tap on the `...` button located on the top bar.
- Tap on the `Post Settings` options.
- Navigate to the Featured Image section and tap on `Set Featured Image`.
- Tap on the `Take a Photo` option.
- Select one time.
- Observe that the Optimize Images pop-up is displayed.</details>

<details>
 <summary>Check events ⤵️</summary>

Verify that the new `APP_SETTINGS_OPTIMIZE_IMAGES_POPUP_TAPPED` event is sent when tapping either of the options in the new popup. `on` will be sent if you leave optimization on and `off` if it's turned off via the popup.
</details>

## Regression Notes
1. Potential unintended areas of impact

As we're changing default settings, there's a risk of unintentionally changing settings users have chosen. We're also changing code around the App Settings screen and media upload flows, so there could be unintended impact there.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manually tested each flow we identified and relied on existing tests around media uploads.

3. What automated tests I added (or what prevented me from doing so)

I plan to further analyse existing tests and potentially add new ones in a future PR.

<hr />

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SiobhyB committed:
Update strings associated with optimization popup

The strings have been updated to reflect the popup's change in functionality. It will now be informing users about the fact optimization is enabled by default, rather than the fact it used to be off by default.

SiobhyB committed:
Update RELEASE-NOTES

SiobhyB committed:
Set "optimize image" setting to "on" by default

SiobhyB committed:
Update popup functionality to reflect new default

SiobhyB committed:
Show new promo in editor when adding local images

The advertiseImageOptimization() function will now be called when a user uploads a photo from their device.

SiobhyB committed:
Introduce tracks event to monitor clicks

SiobhyB committed:
Display new popup after upload via media screen

SiobhyB committed:
Merge branch 'trunk' into feature/change-default-image-optimization-setting

SiobhyB committed:
Remove duplicate case in switch statement

Following the last merge from `trunk`, the `STOCK_MEDIA_PICKER_SINGLE_SELECT` case was accidentally duplicated. This PR reverts that mistake.

SiobhyB committed:
Add missing null check

This was accidentally removed when updating from `trunk` in a previous commit.

SiobhyB committed:
Ensure media is uploaded if popup's not displayed

This commit fixes the bug pinpointed here, in which media wasn't uploading if the popup wasn't displayed: https://github.com/wordpress-mobile/WordPress-Android/pull/19556/files#r1386846666

SiobhyB committed:
Revert unnecessary change to quality default

As noted here, the default yields the same result, whether it's 0 or 1. So, a change is not necessary: https://github.com/wordpress-mobile/WordPress-Android/pull/19556/files#r1386712864

SiobhyB committed:
Refactor popup logic for clarity

Following the review here, the logic surrounding the popup's button has been refactored to make it more clear and readable: https://github.com/wordpress-mobile/WordPress-Android/pull/19556#discussion_r1386733843

SiobhyB committed:
Update wording in popup

The wording has been changed following feedback from the editorial team here: https://github.com/wordpress-mobile/WordPress-iOS/pull/21981#issuecomment-1802410807

SiobhyB committed:
Merge pull request #19556 from wordpress-mobile/feature/change-default-image-optimization-setting

Enable image optimization setting by default

SiobhyB committed:
Merge branch 'trunk' into feature/updates-to-optimization-settings

SiobhyB committed:
Merge branch 'trunk' into feature/updates-to-optimization-settings

SiobhyB committed:
Fix positioning of RELEASE-NOTES entry

SiobhyB committed:
Update values for low and medium image quality

To match the default values for image quality for iOS (see: https://github.com/wordpress-mobile/WordPress-iOS/pull/21981), this commit changes the defaults for quality to the following:

- Low = 70%
- Medium = 80%

SiobhyB committed:
Update string names for accuracy

Following the last commit, the numbers in the string names no longer reflect reality. As such, they're updated in this commit to reflect the level of quality they represent.

SiobhyB committed:
Update default max size to 2000 for optimized images

SiobhyB committed:
Reduce comment length for to pass linter rule

SiobhyB committed:
Merge pull request #19648 from wordpress-mobile/feature/changes-to-default-resolution

Update default max size to 2000 for optimized images

SiobhyB committed:
Merge branch 'feature/updates-to-optimization-settings' into feature/update-optimization-values

SiobhyB committed:
Add check to ensure quality value is valid

Addresses bug reported here: https://github.com/wordpress-mobile/WordPress-Android/pull/19644/files#r1402179477

As we've made changes to the values in the quality array, there may be times when users have previously chosen values that no longer exist. Those cases were not being properly handled in the code.

With this commit, we now check if a quality value exists in the array. If it doesn't, we return the default quality value.

SiobhyB committed:
Import Context for use in class

SiobhyB committed:
Ensure we return default quality int if necessary

In cases when a quality value is invalid, we should explicitly detect that and then return the default quality value. Otherwise, the invalid value will continue to be returned.

Response to the following feedback: https://github.com/wordpress-mobile/WordPress-Android/pull/19644#discussion_r1403179770

SiobhyB committed:
Add extra detail to RELEASE-NOTES

SiobhyB committed:
Merge pull request #19644 from wordpress-mobile/feature/update-optimization-values

Update values associated with "Low" and "Medium" image quality settings

fluiddot commented on WordPress/src/main/res/values-ar/strings.xml:
```diff
@@ -2996,11 +2996,11 @@ Language: ar
-    <string name="media_encoder_quality_90">مرتفع</string>
-    <string name="media_encoder_quality_85">متوسط</string>
-    <string name="media_encoder_quality_80">منخفض</string>
+    <string name="media_encoder_quality_maximum">الحد الأقصى</string>
+    <string name="media_encoder_quality_very_high">مرتفع للغاية</string>
+    <string name="media_encoder_quality_high">مرتفع</string>
+    <string name="media_encoder_quality_medium">متوسط</string>
+    <string name="media_encoder_quality_low">منخفض</string>
```
I wonder if we had to update all localizations. Based on the file history ([example](https://github.com/wordpress-mobile/WordPress-Android/commit/6d575e4ea4964358ec91ee94708145b37ba00d28)), seems these localization files are updated automatically as part of the release process.

SiobhyB replied:
Ah, good to know for the future!

SiobhyB committed:
Rename and reposition tracks event for clarity

"app_settings_optimize_images_popup_changed" has been renamed to "app_settings_optimize_images_popup_tapped" to clarify the action the user's taken. Tapping the popup may not necessarily change the optimization value. This also brings the event in alignment with the naming used on iOS.

The list position of the tracks event has also been changed, so that it's grouped closer to other settings related to image optimization.

fluiddot commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -2974,10 +2974,10 @@
+    <string name="turn_on">Yes, leave on</string>
+    <string name="leave_off">No, turn off</string>
+    <string name="image_optimization_promo_title">Keep optimizing images?</string>
+    <string name="image_optimization_promo_desc">Image optimization shrinks images for faster uploading.\n\This option is enabled by default, but you can change it in the app settings at any time.</string>
```
I noticed one of the break lines (`\n`) only contains `\` after the change:

```suggestion
    <string name="image_optimization_promo_desc">Image optimization shrinks images for faster uploading.\n\nThis option is enabled by default, but you can change it in the app settings at any time.</string>
```

SiobhyB replied:
Committed that change, thanks for spotting!

fluiddot commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -142,12 +143,12 @@ public static void advertiseImageOptimization(final Context context,
+                AnalyticsTracker.track(AnalyticsTracker.Stat.APP_SETTINGS_OPTIMIZE_IMAGES_POPUP_TAPPED,
+                        Collections.singletonMap("option", propertyValue));
+
+                if (which == DialogInterface.BUTTON_NEGATIVE) {
+                    if (AppPrefs.isImageOptimize()) AppPrefs.setImageOptimize(false);
                 }
```
Nitpick: These two conditions could be grouped into a single `if` statement.

SiobhyB replied:
Refactored in https://github.com/wordpress-mobile/WordPress-Android/pull/19581/commits/51ca8afa792ac79662aedf7339ef56280990564a. 🙇‍♀️

fluiddot reviewed (approved):
LGTM 🎊 ! Awesome work @SiobhyB 🏅 !

I've tested the different scenarios described in the PR's description and confirmed the changes behave as expected ✅ . I added some comments, [this one](https://github.com/wordpress-mobile/WordPress-Android/pull/19581#discussion_r1403519469) in specific would be great to address it before merging the PR.

fluiddot commented:
@SiobhyB As a side note and regarding the test section `Default setting should not be changed if it has been set previously`:

Since we've changed the quality values, if the Low value was previously set, the quality displayed after upgrading the app will be Medium. Not sure if users might be confused with this change, I presume that not many users use the lowest quality, hence probably it would be a minor percentage.

SiobhyB committed:
Add additional line break to release notes

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

SiobhyB committed:
Refactor if statement

Refactors two if statements into a single statement, as it was possible to group the two conditions.

Addresses the following feedback: https://github.com/wordpress-mobile/WordPress-Android/pull/19581#discussion_r1403524555

SiobhyB commented:
Thanks @fluiddot! 🙌 

> Since we've changed the quality values, if the Low value was previously set, the quality displayed after upgrading the app will be Medium. Not sure if users might be confused with this change, I presume that not many users use the lowest quality, hence probably it would be a minor percentage.

I agree it may confuse a small percentage of people but, as they'll keep the 80% compression rate, hopefully most won't notice. 🤞 I'll go ahead to merge this PR once the tests pass.
